zenoh-ts/src/pubsub.ts

```diff
@@ -0,0 +1,334 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+// Remote API
+import { RemoteSubscriber, RemotePublisher } from "./remote_api/pubsub";
+
+// API
+import { KeyExpr } from "./key_expr";
+import { IntoZBytes, ZBytes } from "./z_bytes";
+import {
+  CongestionControl,
+  Priority,
+  Reliability,
+  Sample,
+  Sample_from_SampleWS,
+} from "./sample";
+import { Encoding, IntoEncoding } from "./encoding";
+
+// ███████ ██    ██ ██████  ███████  ██████ ██████  ██ ██████  ███████ ██████
+// ██      ██    ██ ██   ██ ██      ██      ██   ██ ██ ██   ██ ██      ██   ██
+// ███████ ██    ██ ██████  ███████ ██      ██████  ██ ██████  █████   ██████
+//      ██ ██    ██ ██   ██      ██ ██      ██   ██ ██ ██   ██ ██      ██   ██
+// ███████  ██████  ██████  ███████  ██████ ██   ██ ██ ██████  ███████ ██   ██
+
+
+/**
+ * Class to represent a Subscriber on Zenoh, 
+ * created via calling `declare_subscriber()` on a `session`
+ */
+
+export class Subscriber {
+  /**
+   * @hidden 
+   */
+  private remote_subscriber: RemoteSubscriber;
+  /**
+   * @hidden 
+   */
+  private callback_subscriber: boolean;
+  /** Finalization registry used for cleanup on drop
+   * @hidden 
+   */
+  static registry: FinalizationRegistry<RemoteSubscriber> = new FinalizationRegistry((r_subscriber: RemoteSubscriber) => r_subscriber.undeclare());
+  /**
+   * @hidden 
+   */
+  dispose() {
+    this.undeclare();
+    Subscriber.registry.unregister(this);
+  }
+  /**
+   * @hidden 
+   */
+  constructor(
+    remote_subscriber: RemoteSubscriber,
+    callback_subscriber: boolean,
+  ) {
+    this.remote_subscriber = remote_subscriber;
+    this.callback_subscriber = callback_subscriber;
+    Subscriber.registry.register(this, remote_subscriber, this)
+  }
+
+  /**
+   * Receives a new message on the subscriber
+   *  note: If subscriber was created with a callback, this recieve will return undefined, 
+   *  as new samples are being sent to the callback.
+   *
+   * @returns Promise<Sample | void>
+   */
+  async receive(): Promise<Sample | void> {
+    if (this.callback_subscriber === true) {
+      console.log("Cannot call `receive()` on Subscriber created with callback:");
+      return;
+    }
+
+    // from SampleWS -> Sample
+    let opt_sample_ws = await this.remote_subscriber.receive();
+    if (opt_sample_ws != undefined) {
+      return Sample_from_SampleWS(opt_sample_ws);
+    } else {
+      console.log("Receieve returned unexpected void from RemoteSubscriber");
+      return;
+    }
+  }
+
+  /**
+   * Undeclares a subscriber on the session
+   *
+   */
+  undeclare() {
+    this.remote_subscriber.undeclare();
+    Subscriber.registry.unregister(this);
+  }
+
+  /**
+   * Create a new subscriber, 
+   * note : This function should never be called directly by the user
+   * please use `declare_subscriber` on a session to create a subscriber
+   * @hidden
+   */
+  static async new(
```

Comment:
I suggest using a Symbol to make this method private: https://www.typescriptlang.org/docs/handbook/symbols.html and making the `constructor` `private`.

Reply:
I believe i have implemented this on https://github.com/eclipse-zenoh/zenoh-ts/pull/5
with the commit https://github.com/eclipse-zenoh/zenoh-ts/pull/5/commits/198f2c4629ecc768b1a9bbd3323da2f5f08e1b46